src/commonMain/kotlin/baaahs/Pinky.kt

```diff
@@ -183,33 +192,13 @@ class Pinky(
         pixelsBySurface[surface] = pixelLocations
     }
 
-    interface BeatProvider {
-        var bpm: Float
-        val beat: Float
-    }
-
-    inner class PinkyBeatProvider(override var bpm: Float) : BeatProvider {
-        private var startTimeMillis = 0L
-        private var beatsPerMeasure = 4
-
-        private val millisPerBeat = 1000 / (bpm / 60)
-
-        override val beat: Float
-            get() {
-                val now = getTimeMillis()
-                return (now - startTimeMillis) / millisPerBeat % beatsPerMeasure
-            }
-
+    inner class PinkyBeatDisplayer(val beatSource: BeatSource) {
         suspend fun run() {
-            startTimeMillis = getTimeMillis()
-
             while (true) {
-                display.beat = beat.toInt()
-
-                val offsetMillis = getTimeMillis() - startTimeMillis
-                val millsPer = millisPerBeat
-                val delayTimeMillis = millsPer - offsetMillis % millsPer
-                delay(delayTimeMillis.toLong())
+                val beatData = beatSource.getBeatData()
+                display.beat = beatData.beatWithinMeasure(clock).toInt() - 1
+                display.bpm = beatData.bpm.toInt() //TODO: show fractions of bpm
+                delay(10)
```

Comment:
Probably oughta schedule this via `requestAnimationFrame()` but nbd.

Reply:
Oh, what would that do?

Reply:
It'll run the code once per frame rather than once per _n_ ms. Small but easy optimization.